solr/core/src/java/org/apache/solr/search/grouping/CommandHandler.java

```diff
@@ -228,12 +226,8 @@ private void searchWithTimeLimiter(Query query,
       collector = MultiCollector.wrap(collector, hitCountCollector);
     }
 
-    if (filter.filter != null) {
-      query = new BooleanQuery.Builder()
-          .add(query, Occur.MUST)
-          .add(filter.filter, Occur.FILTER)
-          .build();
-    }
+    query = QueryUtils.combineQueryAndFilter(query, filter.filter);
```

Comment:
observation: if `query` were null (which presumably it wouldn't be but haven't looked) then `QueryUtils.combineQueryAndFilter` handles that differently now compared to existing logic

Reply:
If the parsed query is null (e.g. was stopwords or something), then the end effect is matching nothing.  You can see this in the former logic by seeing solrQueryWeight getting set to null.  In the new logic, this is done via setting query to MatchNoDocsQuery, and thus "query" variable is non-null.

Reply:
Thanks for the pointer, i see the `MatchNoDocsQuery` now in SolrFeature, hadn't looked before.

Outside of `ltr/SolrFeature` though the `QueryUtils.combineQueryAndFilter` is using `MatchAllDocsQuery`, right?

Reply:
I wouldn't define combineQueryAndFilter by who uses it to reverse engineer how it might work; it has its own javadocs.  It behaves the same no matter who calls it.  Do you think it's defaults are counterintuitive?

BTW I think it's a mistake that Solr's 'q' effectively defaults to semantics of MatchNoDocsQuery.

Reply:
Okay, returning to this I think I now understand the `QueryUtils.combineQueryAndFilter` scenarios better:

* There are two inputs and zero, two or one of them could be null.

  * scoreQuery non-null, filterQuery null:
    * (both old and new code) just use `scoreQuery`.

  * scoreQuery non-null, filterQuery non-null:
    * `BooleanQuery.Builder` is used to combine the two (in both old and new code).

  * scoreQuery null, filterQuery null:
    * In the old code in CommandHandler/SolrIndexSearcher/Grouping/ExpandComponent a value of null would have been used for `scoreQuery`, searching on that isn't gonna end well, is it.
    * In the new code `MatchAllDocsQuery` will be used to return everything.

  * scoreQuery null, filterQuery non-null:
    * In the old code `BooleanQuery.Builder` is used to combine the two, regardless of `scoreQuery` being null and so that search is not going to end well too.
    * In the new code the `filterQuery` wrapped up into a `ConstantScoreQuery` is used to search. In terms of search meaning this is equivalent to the `scoreQuery` being a `MatchAllDocsQuery` i.e. with a score query matching everything only the filter query needs to be searched with.

So in the "scoreQuery null" scenario the new code
* no longer results in a failed search,
* but instead returns something,
* and what is returned is consistent w.r.t. any filter query,
* and/or consistent within itself i.e. a null (score or filter) query is taken to mean 'everything' i.e. `MatchAllDocsQuery`.

So yes, I agree, `MatchAllDocsQuery` is the logical choice.